### DISCUSSION TITLE
Getting an error from ABI when building the indexer

### DISCUSSION THREAD

**Discussion by Walter-Phillips** [FuelLabs/fuel-indexer#1329] [Q&A] [answered]
# Issue

Thanks for opening an issue on the Fuel Indexer project.

_*If you're an internal FuelLabs contributor, please feel free to remove this template when creating issues, as this template is meant for external users and contributors_.

## Prerequisites

- [x] Can you reproduce the problem?
- [x] Do you know which version of the indexer you're using?
- [x] Did you already check for an answer in previous [Discussions](https://github.com/FuelLabs/fuel-indexer/discussions)?

### Description

When calling the `forc index build` command we get this error:

<img width="855" alt="Screenshot 2023-08-29 at 3 05 08 PM" src="https://github.com/FuelLabs/fuel-indexer/assets/34078012/9208062b-94a1-490d-ac9b-ecd781064f46">

### Steps to Reproduce

Please list the steps required to reproduce your issue in detail.

1. Create an indexer project
2. Add the abi attached below to your project and add the path to it in the manifest.yaml
3. Run the `forc index build` command

**Expected behavior:** The indexer builds🤞🏾

**Actual behavior:** We get the error in the attached image:
<img width="855" alt="Screenshot 2023-08-29 at 3 05 08 PM" src="https://github.com/FuelLabs/fuel-indexer/assets/34078012/9208062b-94a1-490d-ac9b-ecd781064f46"> 

**Things we've tried:** 
The contracts for the project are in a workspace so we tried building the contracts with `forc build` instead of `forc build  --json-abi-with-callpaths` and using that abi.

The next thing we tried was moving all the relevant code from a library in a separate Forc project to the same Forc project as the contract that is doing the logging.

**Things we think it could be:**
This wasn't an issue we had before. In our contract, we are using generics and traits. I read something about not having full support for trait bounds in sway. Not sure if that has something to do with it. But contract ABIs that don't have either of these two things haven't caused indexer errors for us in the past.

**Things we will try:**
I'll try rewriting the contracts without traits or generics and see if the error is resolved.

### Versions

Please list what version of indexer 

- `fuel-indexer`: 0.19.5
- `forc-index`: 0.19.5

[order-logger-abi.json.zip](https://github.com/FuelLabs/fuel-indexer/files/12469975/order-logger-abi.json.zip)

**Comment by deekerno** [comment]
@Walter-Phillips: Do you have a link to a repository? I try not to download files if I can help it. 😅 That being said, I think I've seen something similar to this in the past and I'm starting to look into it.

**Comment by Walter-Phillips** [comment]
@deekerno that's fair, I can add you to our repo. The indexer_fix branch should have everything you need. You may have to change the path to the abi on your machine. To build the contracts you can run `just build-contracts`.

**Comment by deekerno** [comment]
@Walter-Phillips Sounds great. You can add @ra0x3 as well so that we're both able to take a look.

**Comment by Walter-Phillips** [comment]
@deekerno @ra0x3 just sent you both an invite, it's in the sail-indexer folder in packages. It's the order-logger ABI that we are using. We appreciate you guys looking into it.🙏🏾

**Comment by deekerno** [comment]
@Walter-Phillips, do you mind trying the following for me?

1. Make sure that you're using Fuel indexer packages with the same version as the `fuel-indexer` binary available on your system. In your case, you would update your `fuel-indexer-utils` dependency as follows: `fuel-indexer-utils = { version = "0.19" }`.
2. Adjust your entity types to use the `@entity` directive. For example:
```
type WalterBlock @entity {
    ...
}
```

We made some adjustments to how users should define entity types, but it wasn't in the documentation yet; coincidentally, I have a PR up for it right now.

I was able to build your indexer following those steps. Let me know if they work for you. 🙏🏽

**Comment by Walter-Phillips** [comment]
@deekerno I tried it with our repo and got the error below. I also created a new indexer project following the steps you gave. I tried it with `fuel-indexer-utils = { version = "0.19" }`. However, on my machine, the version of the fuel-indexer I have is 19.5 so also gave that a try and got the same error. 

To be clear it built on the indexer_fix branch? I've been able to get it to run on main but that doesn't have a lot of our new changes. 

<img width="584" alt="Screenshot 2023-08-30 at 11 39 52 AM" src="https://github.com/FuelLabs/fuel-indexer/assets/34078012/84bcd2db-f292-449d-9b3b-170dd2093cc3">

**Comment by deekerno** [comment]
@Walter-Phillips Ah, I missed the `indexer-fix` branch part. I think I see the issue with this latest error; our code generation that processes the ABI types has included the path separator as part of the identifier, which (as it says) is invalid. I'll start working on a fix and we'll issue a patch release for the v0.19 series, which still supports beta-3.

Also, in your Cargo manifest on your indexer, you're using `fuels = { version = 0.44 }`. That SDK version dropped support for beta-3 as we all prepare for beta-4. You'll need to use `fuels = { version = 0.43 }` to ensure that your indexer is compatible with beta-3.

**Comment by Walter-Phillips** [comment]
@deekerno Most of our up-to-date code is beta-4 compatible. Considering beta-4 is happening soon👀, I'd be happy with just using the latest toolchain and making our indexer beta-4 compatible now, if that's possible. If not happy to use `fuels = { version = 0.43 }`. Thanks again for all the help.

**Comment by deekerno** [comment]
@Walter-Phillips Heard that. You'll just be unable to test indexing beta-4-specific things on a contract until either the beta-4 network is deployed or you deploy a local Fuel node of your own that supports said beta-4 stuff (which I haven't tried on my own, so YMMV). The good news is that when beta-4 is released, you can just upgrade your dependencies to the latest versions and everything should _just work_<sup>TM</sup>; all the necessary changes to the indexer should be transparent to the user.

**Comment by Walter-Phillips** [comment]
@deekerno Makes sense. Please let me know when the patch gets merged if possible. Thanks again.

**Comment by deekerno** [comment]
@Walter-Phillips #1315 will fix the issue that you shared initially. When building your contract with the latest Sway version, I was unable to replicate the invalid identifier error as the ABI JSON output seems to have improved. That being said, we're going to merge this into our `develop` branch which is only compatible with beta-4; do you need us to issue a patch release for the `v0.19` version or are you okay with using the latest indexer version?

**Comment by Walter-Phillips** [comment]
@deekerno All good with using the newest version, would that be 0.20.5?

**Comment by deekerno** [comment, marked as answer]
@Walter-Phillips It would be the next patch release; I can start that process now. I've confirmed that #1315 fixed your original issue. However, it appears that your contract is using generics; unfortunately, the indexer currently does not support Sway generics (due to some SDK difficulties) and will not build when they're present in the ABI.

**Comment by Walter-Phillips** [comment]
@deekerno I can edit the contract to not use generics for now and revert when that fix gets added. Sounds great otherwise. Appreciate you. The quick responses have been greatly appreciated.

**Comment by deekerno** [comment]
@Walter-Phillips No problem. I'm glad that we could fix your issue! v0.20.7 has the fix that you need; setting `fuel-indexer-utils = { version = "0.20" }` and `fuels = { version = "0.46" }` should be all you need to do. Also, you may want to change your `Order` entity to something else so it doesn't clash with your Sway `Order` type. 😄

**Comment by deekerno** [comment]
@Walter-Phillips, I'm going to close this issue. If this wasn't resolved correctly, feel free to reopen and continue the conversation. And as always, feel free to open up an issue if you have any other questions.

Cheers!
